src/group_theory/conjugates.lean

```diff
@@ -0,0 +1,142 @@
+/-
+
+Author: Michael Howes and Scott Morrison, 2019
+
+The definition of conjugates in a group and the definition
+of the normal closure of a subset of a group.
+-/
+import algebra.group
+import group_theory.subgroup
+import tactic.rewrite
+
+variable {α : Type}
+variable [group α]
+
+
+
+/-- Two elements a b of a group are conjugates if there exists a
+c such that c*a*c⁻¹ = b -/
+def is_conjugate (a b : α) : Prop := ∃ c : α, c*a*c⁻¹ = b
+
+/-- Some basic properties of conjugates -/
+@[simp] lemma conj_inv {a b : α} : (b*a*b⁻¹)⁻¹ = b*a⁻¹*b⁻¹ :=
+begin
+  rw [mul_inv_rev _ b⁻¹, mul_inv_rev b _, inv_inv, mul_assoc],
+end
+
+@[simp] lemma conj_mul {a b c : α} : (b*a*b⁻¹)*(b*c*b⁻¹) = b*(a*c)*b⁻¹ :=
+begin
+  assoc_rw inv_mul_cancel_right,
+  repeat {rw mul_assoc},
+end
+
+namespace is_conjugate
+/-- is_conjugate is an equivalence relation -/
+lemma refl {a : α} : is_conjugate a a := exists.intro (1:α) (by simp)
+
+lemma symm_aux {a b c : α} : c*a*c⁻¹ = b → a=c⁻¹*b*c  :=
+λ h, by rw [mul_assoc, ←(eq_inv_mul_of_mul_eq (eq_mul_of_mul_inv_eq h))]
+
+lemma symm {a b : α} : is_conjugate a b → is_conjugate b a :=
+λ ⟨c,h⟩, ⟨c⁻¹, by simpa only [inv_inv] using eq.symm (symm_aux h)⟩
+
+lemma trans_aux {a c d e : α} : d * a * d⁻¹ = e⁻¹ * c * e → e*d*a*(d⁻¹*e⁻¹) = c :=
+λ h,
+begin
+  assoc_rw h,
+  simp,
+end
+
+lemma trans {a b c : α} : is_conjugate a b → is_conjugate b c → is_conjugate a c :=
+λ ⟨d,w⟩ ⟨e,h⟩,
+begin
+  rw (symm_aux h) at w,
+  exact ⟨e*d, by simpa only [mul_inv_rev] using (trans_aux w)⟩
+end
+
+end is_conjugate
+
+/-- Here the normal closure of a set is defined as the subgroup closure of the
+set of conjugates of the given conjugates -/
+
+variables {s : set α}
+
+/-- Given an element a, conjugates a is the set of conjugates -/
+def conjugates (a : α) : set α := is_conjugate a
+
+lemma self_mem_conjugates {a : α} : a ∈ conjugates a := is_conjugate.refl
+
+/-- Given a set s, conjugates_of_set s is the set of all conjugates of
+the elements of s. -/
+def conjugates_of_set (s : set α) : set α :=
+  λ a : α, ∃ x ∈ s, a ∈ conjugates x
+
+theorem mem_conjugates {x : α} : x ∈ s → x ∈ conjugates_of_set s :=
+  λ h, Exists.intro x (Exists.intro h self_mem_conjugates)
+
+theorem subset_conjugates_of_set : s ⊆ conjugates_of_set s :=
+  λ (x : α) (h : x ∈ s), mem_conjugates h
+
+theorem conjugates_of_set_mono {s t : set α} (h : s ⊆ t) : conjugates_of_set s ⊆ conjugates_of_set t :=
+  λ a ⟨b,⟨H,w⟩⟩, ⟨b, ⟨set.mem_powerset h H, w⟩⟩
+
+theorem conjugates_of_set_subset {s t : set α} [normal_subgroup t] (h : s ⊆ t) :
+    conjugates_of_set s ⊆ t :=
+λ a ⟨b,⟨H,⟨c,w⟩⟩⟩,
+begin
+  have w' : c*b*c⁻¹ ∈ t := normal_subgroup.normal b (set.mem_powerset h H) c,
+  rwa ←w,
+end
+
+open group
+
+/-- Defining the normal_closure as described above. -/
```

Comment:
You should write the doc comment so it can be read in isolation (it is usually called up via mousing over the definition in vscode).

Reply:
I'm not sure how to do this.

Reply:
@Michael-Howes Mario just means instead of writing "as described above.", just write again what was described above. That way it will appear as a tool tip.

Reply:
Ideally, there would be a big comment at the top of the file, explaining what happens in the file, and any interesting design considerations.

Then every definition or lemma would have a `doc-comment`, using the `/-- comment -/` syntax (notice the double hyphen at the beginning), which will automatically appear in the pop-up window when someone hovers over the name in another file. This doc comment must be readable in isolation, and basically should answer "what is this symbol, anyway?"

If you want to add "directions to the reader" like you're doing here, that are only relevant to someone actually browsing this file, just put them as normal comments: either `--` at the beginning of the line, or `/- comment -/` for multi-line comments (notice the single hyphen).

Reply:
I've added `doc-comments` to all of the main definitions and results.